test/utils.py

```diff
@@ -0,0 +1,373 @@
+    send_my_pid.set(os.getpid())  # Send it so main process knows what to kill
+    return fn(*args, **kwargs)  # No exception handling - pool does it for us
+
+
+def _pool_worker_2(fn, args, kwargs, debug=0):
+    if debug >= 1:
+        print(f"Worker process running on pid={os.getpid()}")
+    if debug >= 2:
+        time.sleep(15)
+        print('After sleep')
+    return fn(*args, **kwargs)  # No exception handling - pool does it for us
+
+
+async def _run_with_timeout_async_pool(pool: SimpleProcessPool, timeout: float, target,
+                                       args=(), kwargs=None, debug=0):
+    try:
+        return await pool.apply(_pool_worker_2, (target, args, kwargs, debug),
+                                {}, timeout)
+    except TimeoutError as e:
+        raise TimeoutError(f"Function took too long "
+                           f"(exceeded timeout of {timeout}s)") from e
+    except Exception as e:
+        raise mp.ProcessError("Error in process") from e
+
+
+@dataclass
+class _Task:
+    inputs: tuple[int, Callable, tuple, dict[str, ...]]
+    output: tuple[int, bool, object | Exception] = None
+
+
+class SimpleProcessPool:
+    def __init__(self, processes: int = None):
+        self.n_processes = processes or os.cpu_count() or 4
+        self.processes: list[_ProcessWrapper] = self._create_processes()
+        self.key = 0
+        self.tasks: dict[int, _Task] = {}
+
+    def _create_processes(self):
+        return [_ProcessWrapper(self, i)
+                for i in range(self.n_processes)]
+
+    def _kill_and_restart(self, i: int):
+        self.processes[i].p.kill()
+        self.processes[i] = _ProcessWrapper(self, i)
+
+    async def _wait_for_empty_process(self, interval: float = 0):
+        while True:
+            for i, p in enumerate(self.processes):
+                self._update_process(i)
+                if p.is_waiting():
+                    return i, p
+            await asyncio.sleep(interval)
+
+    def _update_process(self, i: int):
+        p = self.processes[i]
+        if p.is_dead():
+            self._kill_and_restart(i)
+        p.update_output_status()
+
+    def _update_all_processes(self):
+        for i, _ in enumerate(self.processes):
+            self._update_process(i)
+
+    async def apply(self, fn, args, kwargs, timeout: float = None, interval: float = 0):
+        key = self.key
+        self.key += 1
+        task = self.tasks[key] = _Task((key, fn, args, kwargs))
+        # Only submit one task at a time, avoids being killed
+        # because something else timed out, etc.
+        idx, p = await self._wait_for_empty_process(interval)
+        p.submit(task)
+        start = time.perf_counter()
+        while self.tasks[key].output is None:
+            if timeout and time.perf_counter() > start + timeout:
+                self._kill_and_restart(idx)
+                del p  # I don't feel the need to keep that in tb_frame.f_locals forever
```

Comment:
Is this really needed? This should only happen in exceptional cases and it might be useful for debugging those.